Validator monitoring

arnetheduck opened the pull request:
Validator monitoring based on and mostly compatible with the
implementation in Lighthouse - tracks additional logs and metrics for
specified validators so as to stay on top on performance.

The implementation works more or less the following way:
* Validator pubkeys are singled out for monitoring - these can be
running on the node or not
* For every action that the validator takes, we record steps in the
process such as messages being seen on the network or published in the
API
* When the dust settles at the end of an epoch, we report the
information from one epoch before that, which coincides with the
balances being updated - this is a tradeoff between being correct
(waiting for finalization) and providing relevant information in a
timely manner)

TODO:
- [x] completed sync committee monitoring
- [x] altair rewards
- [x] docs

zah commented on beacon_chain/validators/validator_duties.nim:
```diff
@@ -483,52 +483,27 @@ proc proposeSignedBlock*(node: BeaconNode,
+      MsgSource.api, wallTime, blck)
+
+    if newBlockRef.isErr:
+      withBlck(newBlock):
```
You are already within a `withBlock` block here, so this extra one is unnecessary.

arnetheduck commented:
Epoch end messages:
```
INF 2021-11-17 20:53:42.108+01:00 Previous epoch attestation included        topics="chaindag" timely_source=true timely_target=true timely_head=true epoch=81972 validator=b04395ff
INF 2021-11-17 20:53:42.108+01:00 Previous epoch attestation included        topics="chaindag" timely_source=true timely_target=true timely_head=false epoch=81972 validator=b173ac90
NOT 2021-11-17 20:53:42.108+01:00 Attestation failed to match head           topics="chaindag" epoch=81972 validator=b173ac90
INF 2021-11-17 20:53:42.108+01:00 Previous epoch attestation included        topics="chaindag" timely_source=true timely_target=true timely_head=true epoch=81972 validator=b869a146
INF 2021-11-17 20:53:42.108+01:00 Previous epoch attestation included        topics="chaindag" timely_source=true timely_target=true timely_head=true epoch=81972 validator=b8a1fc69
INF 2021-11-17 20:53:42.108+01:00 Previous epoch attestation included        topics="chaindag" timely_source=true timely_target=true timely_head=true epoch=81972 validator=b1fa2598
```

tersec commented on beacon_chain/rpc/rest_validator_api.nim:
```diff
@@ -512,6 +518,12 @@ proc installValidatorApiHandlers*(router: var RestRouter, node: BeaconNode) =
             return RestApiResponse.jsonError(Http400,
                                              InvalidValidatorIndexValueError)
+          let validator_pubkey = getStateField(
+            node.dag.headState.data, validators)[item.validator_index].pubkey
```
This is presumably invalidating that `HashList` cache for no particular reason.

arnetheduck replied:
67fa1bec3ee7fae5132b089ed2b4c879eea916ad

tersec commented on beacon_chain/validators/validator_monitor.nim:
```diff
@@ -0,0 +1,746 @@
+    validator_monitor_unaggregated_attestation_delay_seconds.observe(
+      delay.toFloatSeconds(), [$src, metricId])
+
+    info "Attestation seen",
```
These `info`-level logs ensure that the validator monitoring, when enabled, creates nearly unavoidably verbose logs.

arnetheduck replied:
well, that's kind of .. the point :) One downside is that we double-log things effectively - both "sent" and "seen" - it may be that we can revisit some of this duplication should we enable the feature by default.

The way to avoid the verbose logs is to set the log level to `NOT`.

zah commented on beacon_chain/gossip_processing/eth2_processor.nim:
```diff
@@ -473,9 +496,15 @@ proc syncCommitteeContributionValidator*(
+      contributionAndProof, v.get()[0])
+
+    self.validatorMonitor[].registerSyncContribution(
+      src, wallTime, contributionAndProof, v.get()[1])
```
How about using named fields here?

arnetheduck replied:
I don't care/mind either way, but the error return is also an anonymous tuple (called `ValidationError`) so I simply stayed consistent with that - if we change, we should change that too (perhaps in a separate PR)

zah commented on beacon_chain/gossip_processing/gossip_validation.nim:
```diff
@@ -855,10 +856,10 @@ proc validateSignedContributionAndProof*(
-      syncCommitteeSlot,
-      committeeIdx,
-      msg.message.contribution.aggregation_bits):
+  let participants = toSeq(dag.syncCommitteeParticipants(
```
Both this copy and the returned copy can be avoided once we have access to view types in the future. Leaving a TODO comment in the code would be appropriate.

arnetheduck replied:
there are _a lot_ of places that would benefit from avoiding copies - I've considered adding todo:s everywhere but refrained from doing so because often, a view type is not needed (an iterator can be used instead - I've already replaced some of your "view todo":s in that way) and if ever we do get access to view types that wrok, we'll need a tool to discover where it is relevant to optimize - in the meantime, I feel they clutter the code unnecessarily, but if you still think they're useful, I can add them here.

arnetheduck replied:
basically, I've been using vtune to find the ones that matter - for the rest, well, whether it's a view type or not is secondary to how well the code reads and what compiler bugs and unfamiliar features one must risk to get there

zah replied:
I think a TODO comment here would be more useful than the typical trivial opportunities in other places because the required refactoring a bit more involved. Instead of returning the sync sub-committee members, the validation function would return the sub-committee ID. The caller would use a view-types based API to get the list of the members without introducing allocations and/or copying.

arnetheduck replied:
Right, I'm not sure that would fly, as an approach, specially as long as we're using the head committee trick - it's quite fragile (aka I've fixed many such bugs) in the sense that the underlying sync committee might change in an accidental async call somewhere between validation and expansion - I can see the case for returning some lent view that the compiler guarantees to stay fresh. Here, the idea is that we validate the data and use it directly to compute the list then we export the outcome of that computation.

This was one of the downsides of separating validation from updates - I've considered an alternative design where the pools do the validation as part of adding the message - this has other problems, because it doesn't separate presumably fast gossip validation from slow pool mutation. So many choices.

arnetheduck replied:
b0162f73a9035afe61b04038db89c942a88d4e20

zah commented on beacon_chain/rpc/rest_validator_api.nim:
```diff
@@ -486,6 +486,12 @@ proc installValidatorApiHandlers*(router: var RestRouter, node: BeaconNode) =
         request.is_aggregator)
 
+      let validator_pubkey = getStateField(
+        node.dag.headState.data, validators).asSeq()[request.validator_index].pubkey
```
Expensive copy of the `validators` field. Use `withState(..): state.data.validators[request.validator_index]` instead.

arnetheduck replied:
as noted in the other PR, there's no expensive copy of `validators` here - there is a somewhat expensive copy of `Validator` however that would need `lent` to avoid.

Here's how the code expands:
```
									T78_ = (tyObject_HashList__EbZJBpzKVMIaHsYOwWyCGQ *)0;
									switch ((*(*(*(*(*colonenvP_).colonup_).colonup_).node1).dag).headState.data.kind)
									{
									case ((tyEnum_BeaconStateFork__Dw9aV5NV0W0A7EnWLN5Ji9aw)2):
									{
										T78_ = (&(*(*(*(*(*colonenvP_).colonup_).colonup_).node1).dag).headState.data.mergeData.data.validators);
									}
									break;
									case ((tyEnum_BeaconStateFork__Dw9aV5NV0W0A7EnWLN5Ji9aw)1):
									{
										T78_ = (&(*(*(*(*(*colonenvP_).colonup_).colonup_).node1).dag).headState.data.altairData.data.validators);
									}
									break;
									case ((tyEnum_BeaconStateFork__Dw9aV5NV0W0A7EnWLN5Ji9aw)0):
									{
										T78_ = (&(*(*(*(*(*colonenvP_).colonup_).colonup_).node1).dag).headState.data.phase0Data.data.validators);
									}
									break;
									}
									(*colonenvP_).validator_pubkey16 = (*T78_).data->data[(*colonenvP_).request7.validator_index].pubkey;
									addAutoMonitor__Wv11X89aUgvCIKIDyjDFXRA((*(*(*(*colonenvP_).colonup_).colonup_).node1).validatorMonitor, (&(*colonenvP_).validator_pubkey16), (*colonenvP_).request7.validator_index);
```

I'd generally say the biggest issue here is that `validator_pubkey` is included in the closure environment - it's not a problem here, but it is a big problem in `async` in general, that variables that "could" be local end up spuriously being part of the environment - this increases memory consumption significantly.

zah replied:
Okay, I hadn't noticed this improvement, sorry for the wasted time.

zah commented on beacon_chain/rpc/rest_validator_api.nim:
```diff
@@ -512,6 +518,12 @@ proc installValidatorApiHandlers*(router: var RestRouter, node: BeaconNode) =
             return RestApiResponse.jsonError(Http400,
                                              InvalidValidatorIndexValueError)
+          let validator_pubkey = getStateField(
+            node.dag.headState.data, validators).asSeq()[item.validator_index].pubkey
```
Ditto

arnetheduck replied:
Ditto

zah commented on beacon_chain/spec/state_transition_epoch.nim:
```diff
@@ -648,6 +648,15 @@ iterator get_flag_index_deltas*(
       if is_unslashed_participating_index(
           state, flag_index, previous_epoch, vidx):
+
+        let pflag = case flag_index
```
This looks like a reusable function.

arnetheduck replied:
it does indeed, but there's nowhere else to use it, so yagni.

a bigger refactoring would perhaps diverge from the spec in how `flag_index` is represented throughout, which perhaps would net a higher readability reward - the way it's done current is _quite_ convoluted.

zah commented on beacon_chain/validators/validator_monitor.nim:
```diff
@@ -0,0 +1,746 @@
+template withEpochSummary(
+  self: var ValidatorMonitor, monitor: var MonitoredValidator,
+  epochParam: Epoch, body: untyped) =
+  mixin summaryIdx
```
Why is this mixed in? You don't want the user to override it in the caller scope.

arnetheduck replied:
I had a random compile issue and hoped this would fix it, why else would one use mixin? :)

it turned out to be the following, very annoying behavior: if there's a symbol called `epochSummary` in another module - anywhere - it will be matched over the local `epochSummary` below, when expanding `body` - quite annoying, any ideas what to do about it? I'd like `body` to prefer the "local" template.

zah replied:
Yes, this is a nasty effect of the Nim generics pre-pass (which I've tried to convince @Araq to remove). 

Nim wants to assign symbol candidates to all identifiers in the body of generics and templates before the said generics and templates are instantiated. If there is only a single candidate for a particular symbol, it's bound early before the code has a chance to consider such injected symbols. So, a work-around would be to introduce a dummy overload of `epochSummary` next to the top-level one that was inappropriately selected. This would force Nim to keep the set of pre-pass matches as an "open sym choice" and your injected definition should be discovered later during the actual instantiation.

The use of `mixin summaryIdx` should not help for this problem in any way, so I would still suggest removing it.

arnetheduck replied:
ec860cb1eb9728e614b6413df9c9962b7fbda9d1

zah commented on beacon_chain/validators/validator_monitor.nim:
```diff
@@ -0,0 +1,746 @@
+      syncCommitteeMessage = shortLog(sync_committee_message.beacon_block_root),
+      src, epoch = slot.epoch, validator = id
+
+    self.withEpochSummary(monitor, slot.epoch):
```
The same sync committee message can reach more than once here. As far as I can tell, this would lead to over-counting in the metrics.

arnetheduck replied:
This is in line with the lighthouse implementation - I guess this is why we record a `minimum` per-epoch delay.

That said, I'm unsure how it would happen exactly - sync committee messages _should_ be deduplicated by gossipsub (this function is called from eth2_processor for valid messages only), unless a single validator has two spots in the committee in which case it should be recorded twice, no?

zah replied:
There are multiple layers of deduplication based on time eviction caches (one in gossipsub and another that is reset once per slot in the sync committee msg pool). Nevertheless, we know that gossip messages "echo" in the network and they end up being processed more than once. The bug with invalid contributions was happening because some sync committee messages were being processed more than once by a similar function in the sync committee msg pool.

zah replied:
So, I guess the idea of the metric is to use the fact that we are seeing a message more than once as a signal of some sort that the message wes broadcasted well in the network. Seems a bit dubious logic for the stated arguments, but alright.

arnetheduck replied:
> broadcasted well in the network

indeed - "it got this far"

arnetheduck replied:
> Nevertheless, we know that gossip messages 

sync committee messages have a pretty short lifespan, no? should we tune the gossip cache timeout?

arnetheduck replied:
b3545ac202ee8c6ab2b4ba83570d21463d41c101

zah commented on beacon_chain/validators/validator_monitor.nim:
```diff
@@ -0,0 +1,746 @@
+      src, epoch = slot.epoch, validator = id
+
+    self.withEpochSummary(monitor, slot.epoch):
+      epochSummary.sync_contributions += 1
```
Ditto

arnetheduck replied:
sync contributions, like aggregate inclusions, are indeed double-counted, in line with the lighthouse implementation - this is a quality metric of sorts: how many others included my attestation / syncsig? if it's not "all aggregators", something is wrong.

zah replied:
Overcounting could still mislead your though. You may count the same contribution multiple times and then conclude that all aggregators have seen your sync committee messages which is not necessary the case. So, the usefulness is limited to "higher counts are better than lower counts".

zah commented on beacon_chain/validators/validator_monitor.nim:
```diff
@@ -0,0 +1,746 @@
+      head = beacon_block_root, slot = slot, validator = id
+
+    self.withEpochSummary(monitor, slot.epoch):
+      epochSummary.sync_signature_block_inclusions += 1
```
What if the signature ends up in multiple orphaned blocks? The resulting metrics may be misleading to the user.

arnetheduck replied:
yes, this is misleading indeed, but also in line with the lighthouse implementation - the same thing can happen for all the metrics because of reorgs, but is perhaps more probable for sync committees - I'm not sure what a better metric would be, that is viable from a state transition perspective (ie that doesn't require examining a complex out-of-state history) - any good ideas?

zah replied:
Well, why do we need to replicate the Lighthouse implemetnation if has flaws like these. Our users would probably lose trust in these metrics if they see that they've lost money at a particular epoch, but the metrics report everything to be fine.

We can perhaps have another set of metrics that report the precise performance as soon as the chain finalizes.

arnetheduck replied:
> replicate

The feature is here so that dashboards and alerts become portable, by and large, so I reasoned we want to keep it pretty close.

Also, if we wait until finalization, we need to consider the additional delay that happens in reporting, we need to keep a dynamic number of epochs "open", multiple branches of history as well as the stampede of reports on finalization - waiting is not without its own issues.

arnetheduck replied:
b3545ac202ee8c6ab2b4ba83570d21463d41c101

unixpi commented on docs/the_nimbus_book/src/validator-monitor.md:
```diff
@@ -0,0 +1,61 @@
+# Validator monitoring
+
+The validator monitoring feature allows tracking the life-cycle and performance of one or more validators in detail. Monitoring can be done for any validator with slightly more detail for validators that are running through the same beacon node.
```
```suggestion
The validator monitoring feature allows for tracking the life-cycle and performance of one or more validators in detail. Monitoring can be carried out for any validator, with slightly more detail for validators that are running through the same beacon node.
```

unixpi commented on docs/the_nimbus_book/src/validator-monitor.md:
```diff
@@ -0,0 +1,61 @@
+
+The validator monitoring feature allows tracking the life-cycle and performance of one or more validators in detail. Monitoring can be done for any validator with slightly more detail for validators that are running through the same beacon node.
+
+Every time the validator performs an duty, the duty is recorded and the monitor will keep track of events related to getting rewarded for having performed it:
```
```suggestion
Every time the validator performs a duty, the duty is recorded and the monitor keeps track of the reward-related events for having performed it. For example:
```

unixpi commented on docs/the_nimbus_book/src/validator-monitor.md:
```diff
@@ -0,0 +1,61 @@
+* When attesting, the attestation being added to an aggregate, then block, then being given a reward in the state
+* When performing sync committee duties, likewise
+
+Validator actions can be traced either through logging, or comprehensive metrics that allow creating alerts in monitoring tools. The metrics are based on the same feature in [Lighthouse](https://lighthouse-book.sigmaprime.io/validator-monitoring.html), thus dashboards and alerts can be used with either client.
```
```suggestion
Validator actions can be traced either through logging, or comprehensive metrics that allow for creating alerts in monitoring tools. The metrics are based on the same feature in [Lighthouse](https://lighthouse-book.sigmaprime.io/validator-monitoring.html), thus dashboards and alerts can be used with either client.
```

unixpi commented on docs/the_nimbus_book/src/validator-monitor.md:
```diff
@@ -0,0 +1,61 @@
+
+Every time the validator performs an duty, the duty is recorded and the monitor will keep track of events related to getting rewarded for having performed it:
+
+* When attesting, the attestation being added to an aggregate, then block, then being given a reward in the state
```
```suggestion
* When attesting, the attestation is added to an aggregate, then a block, before a reward is applied to the state
```

unixpi commented on docs/the_nimbus_book/src/validator-monitor.md:
```diff
@@ -0,0 +1,61 @@
+
+When a validator performs a duty, such as signing an attestation or a sync committee message, this is broadcast to the network. Other nodes pick it up and package the message into an aggregate and later a block. The block is included in the canonical chain and a reward is given two epochs (~13 minutes) later.
+
+The monitor tracks these actions and will log each step at the INF level. If any step is missed, a NOT log is shown instead.
```
```suggestion
The monitor tracks these actions and will log each step at the `INF` level. If any step is missed, a `NOT` log is shown instead.
```

unixpi commented on docs/the_nimbus_book/src/validator-monitor.md:
```diff
@@ -0,0 +1,61 @@
+
+The monitor tracks these actions and will log each step at the INF level. If any step is missed, a NOT log is shown instead.
+
+A typical lifecycle of an attestation might look something like the following:
```
```suggestion
The typical lifecycle of an attestation might look something like the following:
```

unixpi commented on docs/the_nimbus_book/src/validator-monitor.md:
```diff
@@ -0,0 +1,61 @@
+
+The lifecycle of a particular message can be traced by following the `epoch=.... validator=...` fields in the message.
+
+Failures at any point are recorder at a higher logging level, such as `NOT`(ice):
```
```suggestion
Failures at any point are recorded at a higher logging level, such as `NOT`(ice):
```

unixpi commented on docs/the_nimbus_book/src/validator-monitor.md:
````diff
@@ -0,0 +1,61 @@
+NOT 2021-11-17 20:53:42.108+01:00 Attestation failed to match head           topics="chaindag" epoch=81972 validator=...
+```
+
+:warning: It should be noted that metrics are tracked for the current history - in the case of a reorg on the chain - in particular a deep reorg - no attempt is made to revisit previously reported values. In the case that finality is delayed, the risk of stale metrics increases.
````
```suggestion
> :warning: It should be noted that metrics are tracked for the current history - in the case of a reorg on the chain - in particular a deep reorg - no attempt is made to revisit previously reported values. In the case that finality is delayed, the risk of stale metrics increases.
```

unixpi commented on docs/the_nimbus_book/src/validator-monitor.md:
```diff
@@ -0,0 +1,61 @@
+
+:warning: It should be noted that metrics are tracked for the current history - in the case of a reorg on the chain - in particular a deep reorg - no attempt is made to revisit previously reported values. In the case that finality is delayed, the risk of stale metrics increases.
+
+Likewise, many metrics, such as aggregation includion, reflect conditions on the network - it may happen that the same message is counted more than once under certain conditions.
```
```suggestion
Likewise, many metrics, such as aggregation inclusion, reflect conditions on the network - it may happen that the same message is counted more than once under certain conditions.
```

unixpi reviewed (commented):
minor edits, mainly style

arnetheduck committed:
Validator monitoring

Validator monitoring based on and mostly compatible with the
implementation in Lighthouse - tracks additional logs and metrics for
specified validators so as to stay on top on performance.

The implementation works more or less the following way:
* Validator pubkeys are singled out for monitoring - these can be
running on the node or not
* For every action that the validator takes, we record steps in the
process such as messages being seen on the network or published in the
API
* When the dust settles at the end of an epoch, we report the
information from one epoch before that, which coincides with the
balances being updated - this is a tradeoff between being correct
(waiting for finalization) and providing relevant information in a
timely manner)
